include/ddc/kernels/splines/splines_linear_problem_2x2_blocks.hpp

```diff
@@ -0,0 +1,258 @@
+     *
+     * A = |   Q    |             0             |  | I | Q^-1*gamma |
+     *     | lambda | delta - lambda*Q^-1*gamma |  | 0 |      I     |
+     *
+     * So we perform the factorization inplace to store only the relevant blocks in the matrix (while factorizing
+     * the blocks themselves if necessary):
+     *
+     * |   Q    |         Q^-1*gamma        |
+     * | lambda | delta - lambda*Q^-1*gamma |
+     */
+    void setup_solver() override
+    {
+        // Setup the top-left solver
+        m_top_left_block->setup_solver();
+
+        // Compute Q^-1*gamma in top-right block
+        m_top_right_block.modify_host();
+        m_top_right_block.sync_device();
+        m_top_left_block->solve(m_top_right_block.d_view);
+        m_top_right_block.modify_device();
+        m_top_right_block.sync_host();
+
+        // Push lambda on device in bottom-left block
+        m_bottom_left_block.modify_host();
+        m_bottom_left_block.sync_device();
+
+        // Compute delta - lambda*Q^-1*gamma in bottom-right block & setup the bottom-right solver
+        compute_schur_complement();
+        m_bottom_right_block->setup_solver();
+    }
+
+    /**
+     * @brief Compute y <- y - LinOp*x or y <- y - LinOp^t*x.
+     *
+     * [SHOULD BE PRIVATE (GPU programming limitation)]
+     *
+     * @param x
+     * @param y
+     * @param LinOp
+     * @param transpose
+     */
+    void gemv_minus1_1(
+            MultiRHS const x,
+            MultiRHS const y,
+            Kokkos::View<double**, Kokkos::LayoutRight, typename ExecSpace::memory_space> const
+                    LinOp,
+            bool const transpose = false) const
+    {
+        assert(!transpose && LinOp.extent(0) == y.extent(0)
+               || transpose && LinOp.extent(1) == y.extent(0));
+        assert(!transpose && LinOp.extent(1) == x.extent(0)
+               || transpose && LinOp.extent(0) == x.extent(0));
+        assert(x.extent(1) == y.extent(1));
+
+        Kokkos::parallel_for(
+                "gemv_minus1_1",
+                Kokkos::TeamPolicy<ExecSpace>(y.extent(0) * y.extent(1), Kokkos::AUTO),
+                KOKKOS_LAMBDA(
+                        const typename Kokkos::TeamPolicy<ExecSpace>::member_type& teamMember) {
+                    const int i = teamMember.league_rank() / y.extent(1);
+                    const int j = teamMember.league_rank() % y.extent(1);
+
+                    double LinOpTimesX = 0.;
+                    Kokkos::parallel_reduce(
+                            Kokkos::TeamThreadRange(teamMember, x.extent(0)),
+                            [&](const int l, double& LinOpTimesX_tmp) {
+                                if (!transpose) {
+                                    LinOpTimesX_tmp += LinOp(i, l) * x(l, j);
+                                } else {
+                                    LinOpTimesX_tmp += LinOp(l, i) * x(l, j);
+                                }
```

Comment:
This condition is called at each iteration even though it is a loop invariant. I am thinking we could either:
1. Place it outside the outer loop to mitigate the cost
2. Define an intermediate LinOpT that would have a LayoutLeft and the dimensions reversed ?

Reply:
So we cannot rely on the compiler to optimize this even if transpose is a const ? I'll go to solution 1. because solution 2. may be annoying if we want to give a try to LayoutLeft matrix blocks.

Reply:
Mmmh maybe the GPU compiler could move the condition in between the two loops. I would not bet that it would move it out of the outer loop. It would mean splitting a GPU kernel in two ? Highly doubt it

Reply:
Yeah ok I agree